### PR TITLE
History page - add Incident Details and fix layout

### PR DESCRIPTION
**Pull request by vittoriaThinkst**
## Proposed changes

Add incident details
Fix layout

Screenshot 2024-04-12 at 11 21 17

https://github.com/thinkst/canarytokens/assets/126554007/0ee67a38-d7d0-4c8c-8b4d-7c2c23ccd326

### CODE REVIEW

File: frontend_vue/src/api/main.ts [hunk 1]
```diff
@@ -43,7 +43,7 @@ export function manageToken(params: TokenAuthType) {
     .catch((error) => error.response);
 }
 
-export function downloadToken(params: DownloadTokenType) {
+export function downloadAsset(params: DownloadAssetType) {
```
Comment (mclmax) [inline frontend_vue/src/api/main.ts:46]: why the name change?

File: frontend_vue/src/components/tokens/types.ts [hunk 2]
```diff
@@ -107,38 +107,59 @@ type genericObjectType = {
...
+  'X-Real-Ip': string;
+  'X-Forwarded-For': string;
+  'X-Forwarded-Host': string;
+  Connection: string;
+  'Upgrade-Insecure-Requests': string;
+  'User-Agent': string;
+  Accept: string;
+  'Accept-Encoding': string;
+  'Accept-Language': string;
+};
```
Comment (mclmax) [inline frontend_vue/src/components/tokens/types.ts:141]: i suspect there are going to be lots of other headers here. can we make this just an `Any` dict? What happens when we get headers we don't expect? Breakage?

File: frontend_vue/src/components/ui/IncidentDetails.vue [hunk 3]
```diff
@@ -1,80 +1,145 @@
...
+            <template v-if="isNotEmptyObject(val)">
+              <li
+                v-for="(subval, subkey) in val"
+                :key="subkey"
+                class="break-words"
+              >
+                <span class="block text-xs uppercase text-grey-500">{{
+                  subkey
+                }}</span>
+                <span :class="addClassToBoolean(val)">{{ subval }}</span>
```
Comment (mclmax) [inline frontend_vue/src/components/ui/IncidentDetails.vue:59]: `addClassToBoolean` is a weird name for changing the text colour?

File: frontend_vue/src/components/ui/IncidentDetails.vue [hunk 4]
```diff
@@ -1,80 +1,145 @@
...
+
+console.log(props.hitAlert, 'hitsList');
+
+function isNotEmptyObject(
+  val: boolean | string[] | undefined | HitsType
+): boolean {
+  return typeof val === 'object' && Object.keys(val)?.length > 0;
+}
+
+function addClassToBoolean(val: boolean | HitsType | undefined) {
```
Comment (mclmax) [inline frontend_vue/src/components/ui/IncidentDetails.vue:89]: when are we using green vs red text?

Comment (mclmax) [reply frontend_vue/src/components/ui/IncidentDetails.vue:89]: just for bool values... ok... `highlightBoolean` or something?

Comment (mclmax) [reply frontend_vue/src/components/ui/IncidentDetails.vue:89]: `colourBoolean`?

File: frontend_vue/src/components/ui/IncidentDetails.vue [hunk 5]
```diff
@@ -1,80 +1,145 @@
...
+  return Object.entries(obj).reduce((acc, [key, value]) => {
+    if (value !== null && value !== undefined) {
+      //@ts-ignore
+      acc[key as keyof HitsType] =
+        typeof value === 'object' ? cleanupObject(value) : value;
+    }
+    console.log(acc);
+    return acc;
+  }, {} as HitsType);
+}
```
Comment (mclmax) [inline frontend_vue/src/components/ui/IncidentDetails.vue]: what are we doing here, just stripping out null values?

Comment (mclmax) [reply frontend_vue/src/components/ui/IncidentDetails.vue]: are we sure we won't need those somewhere?

File: frontend_vue/src/components/ui/IncidentDetails.vue [hunk 6]
```diff
@@ -1,80 +1,145 @@
...
+    ...(isNotEmptyObject(requestHeaders) && {
+      'Request Headers': requestHeaders,
+    }),
+    ...(isNotEmptyObject(additionalInfo) && {
+      'Additional Info': additionalInfo,
+    }),
+    ...(isNotEmptyObject(requestArgs) && {
+      'Request Args': requestArgs,
+    }),
+  };
```
Comment (mclmax) [inline frontend_vue/src/components/ui/IncidentDetails.vue:139]: this is super hard to read, can we make it clearer that it's returning a dict with so and so keys / values

**Review by mclmax — CHANGES_REQUESTED** [review]

File: frontend_vue/src/api/main.ts (hunk 1 above)
Comment (vittoriaThinkst) [reply frontend_vue/src/api/main.ts:46]: This endpoint is not used only for tokens: in this diff I use the same endpoint to download the alert list CSV or JSON

File: frontend_vue/src/components/tokens/types.ts (hunk 2 above)
Comment (vittoriaThinkst) [reply frontend_vue/src/components/tokens/types.ts:141]: Yeah we can make it all generic, like I did for additional_info and request args
`export type HitsType = {
  time_of_hit: number;
  src_ip: string;
  geo_info: GeoInfoType;
  is_tor_relay: boolean;
  input_channel: string;
  src_data: genericObjectType;
  useragent: string;
  token_type: string;
  request_headers: RequestHeadersType;
  additional_info: {
    [k: string]: {
      [k: string]: string[];
    };
  };
  request_args: {
    [k: string]: {
      [k: string]: string[];
    };
  };
};`

Nothing happens if the API returns something different, Typescript checks only at compile time

File: frontend_vue/src/components/ui/IncidentDetails.vue (hunk 3 above)
Comment (vittoriaThinkst) [reply frontend_vue/src/components/ui/IncidentDetails.vue:59]: yeah you're right, I will change it  🥲

File: frontend_vue/src/components/ui/IncidentDetails.vue (hunk 5 above)
Comment (vittoriaThinkst) [reply frontend_vue/src/components/ui/IncidentDetails.vue]: uhm I'm not sure about it, do you have a scenario in mind, where a null value is important for the user? 🧐
Thing is we get a lot of empty obj/null values from the backend. I'm not sure we want to show all of them to the user

File: frontend_vue/src/components/ui/IncidentDetails.vue (hunk 4 above)
Comment (vittoriaThinkst) [reply frontend_vue/src/components/ui/IncidentDetails.vue:89]: yeah, makes sense, I will change it to colourBoolean 🌈

File: frontend_vue/src/components/ui/IncidentDetails.vue (hunk 6 above)
Comment (vittoriaThinkst) [reply frontend_vue/src/components/ui/IncidentDetails.vue:139]: do you mean by adding a comment? or splitting the code somehow?

File: frontend_vue/src/api/main.ts (hunk 1 above)
Comment (mclmax) [reply frontend_vue/src/api/main.ts:46]: maybe we just rename the function to `download` then? we can keep the type?

File: frontend_vue/src/components/tokens/types.ts (hunk 2 above)
Comment (mclmax) [reply frontend_vue/src/components/tokens/types.ts:141]: Yeah i'm not sure it's useful keeping a list of this then as the headers object is going to change depending on the requests

File: frontend_vue/src/components/ui/IncidentDetails.vue (hunk 6 above)
Comment (mclmax) [reply frontend_vue/src/components/ui/IncidentDetails.vue:139]: so i'm not sure what this thing is returning at a glance

Comment (mclmax) [reply frontend_vue/src/components/ui/IncidentDetails.vue:139]: is it just an object with keys including things like `'Is Tor Rely': props.hitAlert.is_tor_relay,`. i'm not sure what this `...(isNotEmptyObject(geoInfo) && { 'Geo Info': geoInfo })` is doing?

File: frontend_vue/src/components/ui/IncidentDetails.vue (hunk 5 above)
Comment (mclmax) [reply frontend_vue/src/components/ui/IncidentDetails.vue]: so i'm going to spin this back to you and as are you sure we don't display empty things anywhere in the triggered alert list?

File: frontend_vue/src/components/ui/IncidentDetails.vue (hunk 6 above)
Comment (vittoriaThinkst) [reply frontend_vue/src/components/ui/IncidentDetails.vue:139]: yeah it's not quite straight-forward, let me try if I can make it look better

**Review by mclmax — APPROVED** [review]

**Pull request merged**